### DISCUSSION TITLE
Edge_index construction with multiple graphs for graph classification

### DISCUSSION THREAD

**Discussion by evantianzg** [pyg-team/pytorch_geometric#9150] [Q&A]
Hi,

I am performing a graph classification task using GCN. I successfully constructed my own dataset, but there was an error when running the code. The following is the complete code and error information.

![1](https://github.com/pyg-team/pytorch_geometric/assets/22927915/4b35650e-d83c-4260-bf68-6af7d0b9cb17)

I am not sure how this error occurs. Will the isolated nodes cause this error? (there are no isolated nodes in my data in normal cases, but I found there are isolated nodes in some of graphs in my datasets.)

Any help or suggestions are appreciated. Thanks!

**Comment by rusty1s** [comment]
What happens if you run
```python
for data in dataset:
    data.validate()

**Reply by evantianzg** [reply]
it throws a Value Error "ValueError: 'edge_index' contains larger indices than the number of nodes (2) in 'Data' (found 4)"

So the problem comes from my raw data files?

**Reply by evantianzg** [reply]
@rusty1s  I am sure now that the problem comes from the data. I also checked the `dataset.edge_index.max()` and the `dataset.edge_index.min()`, it shows `tensor(82492)` and `tensor(1)`. The size of `x` obtained by `print(dataset.x.size())` is `torch.Size([41745, 7])`. So the error arises from `num_nodes is not None and self.edge_index.max() >= num_nodes`. 

But the size and shape of my data look all right. So I am not sure if it is the problem from my code in process method or just the raw data. I was wondering if PyG has a source code to process the public dataset like MUTAG. So I can locate the specific problem.

Thanks for your great help.

**Reply by rusty1s** [reply]
Yeah, `edge_index` needs to hold values between [0, num_nodes - 1]. In your case, I assume you write raw IDs to `edge_index`, which is incorrect. You need to map them to a contiguous range before. Our Data from CSV tutorial goes over this.

**Reply by evantianzg** [reply]
Yeah, I understand that. I also read [load_csv](https://github.com/pyg-team/pytorch_geometric/blob/master/docs/source/tutorial/load_csv.rst) tutorial in PyG repository. I found that in this tutorial, the data is processed from the literal raw data (not the so-called raw data under the `raw` directory like MUTAG dataset, which is presented with IDs and digital labels). For example, the nodes and labels in literal raw data may be `names` and `ratings`, but the nodes and labels in so-called raw data under the `raw` directory are all numbers. The so-called raw data under the `raw` directory is actually pre-processed data.

So my question is: what is the best way to convert the literal raw data to a tensor and a data object of PyG? Could I use the pre-processed data to create a data object or just use the literal raw data? The pre-processed data has no IDs, and the literal raw data has IDs. So I suppose it is easier to create a data object from the literal raw data than the pre-processed data?

And also, I use the pre-processed files to create the data objects, so there are no raw IDs in my data. Actually all my data is correctly converted into tensor except edge_index: the value in `edge_index` tensor is inconsistent with the value in my `graph_A.csv` file. I understand that the graphs need to be stacked in a diagonal manner. But I am not sure if I should convert each graph into individual data objects first and then stack them diagonally or just convert all the graphs into an individual object directly. In my code I choose the latter, although it does not work :) 

Any help would be highly appreciated.

**Reply by rusty1s** [reply]
You can basically create a PyG `data` object from any data in any form - you just have to process it accordingly. In general, the `raw` directory in the `dataset` folders just refer to the raw data that we download. In some cases, this might be already pre-processed, in other cases it isn't. That's why the `process` method differs from dataset to dataset. It's hard to give a general answer here how you should do it since this depends on the raw data you have.

**Reply by evantianzg** [reply]
Hi, Sorry for the late reply.

I have fixed the problem with the help of @EdisonLeeeee. PyG is a great tool for graph learning, and the developers, especially you and Edison, are all very nice to help people solve problems.

I am not sure how many rookies like me have asked some simple questions. I just wonder if we could do it better by providing more standard practices. For example, when a rookie like me starts to use PyG, I usually look at the official examples provided, especially to understand what the raw data looks like and how it is processed. On this basis, I suppose a standard practice without limiting other practices should be very helpful since beginners may be trapped in many trivial but still important points. 

This is just my two cents. Thank you for all your great help again.
